### PR TITLE
feat: cache BLOCK_INVALID/BLOCK_VALID status

### PR DESCRIPTION
**Pull request by keroro520**
* feat: Add `BLOCK_VALID` status.
* test: Add test to check the block status after switching forks

### CODE REVIEW

File: sync/src/types.rs [hunk 1]
```diff
@@ -145,7 +145,7 @@ impl PeerState {
     }
 
     // Not use yet
-    pub fn caught_up_sync(&mut self) {
+    pub fn caught_nup_sync(&mut self) {
```
Comment (u2) [inline sync/src/types.rs]: Typo?

Comment (keroro520) [reply sync/src/types.rs]: Thanks!

File: sync/src/block_status.rs [hunk 2]
```diff
@@ -7,6 +7,7 @@ bitflags! {
         const HEADER_VALID            =     0b0000_0000_0000_0001;
         const BLOCK_RECEIVED          =     Self::HEADER_VALID.bits | 0b0000_0000_0000_0010;
         const BLOCK_STORED            =     Self::HEADER_VALID.bits | Self::BLOCK_RECEIVED.bits | 0b0000_0000_0000_1000;
+        const BLOCK_VALID             =     Self::HEADER_VALID.bits | Self::BLOCK_RECEIVED.bits | Self::BLOCK_STORED.bits | 0b0000_0000_0001_0000;
```
Comment (TheWaWaR) [inline sync/src/block_status.rs]: `(1 << 4)` instead of `0b0000_0000_0001_0000` would be more readable?

Comment (keroro520) [reply sync/src/block_status.rs]: Thanks! Updated as you suggested.

And I add another [commit](https://github.com/nervosnetwork/ckb/pull/1237/commits/b4a37519169a1b480bf527ccbf1fd0d9a05e7e4b), PTAL.

**Pull request merged**